fs/xfstests.py

```diff
@@ -0,0 +1,168 @@
+        tests = glob.glob(self.srcdir + '/tests/*/???.out')
+        tests += glob.glob(self.srcdir + '/tests/*/???.out.linux')
+        tests = [t.replace('.linux', '') for t in tests]
+        tests_list = [t[-7:-4] for t in tests if os.path.exists(t[:-4])]
+        tests_list.append('')
+        tests_list.sort()
+        return tests_list
+
+    def _run_sub_test(self, test):
+
+        os.chdir(self.srcdir)
+        output = process.system_output('./check %s' % test,
+                                       ignore_status=True)
+        lines = output.split('\n')
+        result_line = lines[-3]
+        if self.NA_RE.match(result_line):
+            detail_line = lines[-3]
+            match = self.NA_DETAIL_RE.match(detail_line)
+            if match is not None:
+                error_msg = match.groups()[2]
+            else:
+                error_msg = 'Test dependency failed, test not run'
+            raise self.error(error_msg)
+
+        elif self.FAILED_RE.match(result_line):
+            raise self.error('Test error, check debug logs for complete '
+                             'test output')
+
+        elif self.PASSED_RE.match(result_line):
+            return
+
+        else:
+            raise self.error('Could not assert test success or failure, '
+                             'assuming failure. Please check debug logs')
+
+    def _get_groups(self):
+        '''
+        Returns the list of groups known to xfstests
+        By reading the group file and identifying unique mentions of groups
+        '''
+        groups = []
+        for l in open(os.path.join(self.srcdir, 'group')).readlines():
+            m = self.GROUP_TEST_LINE_RE.match(l)
+            if m is not None:
+                groups = m.groups()[1].split()
+                for g in groups:
+                    if g not in groups:
+                        groups.add(g)
+        return groups
+
+    def _get_tests_for_group(self, group):
+        '''
+        Returns the list of tests that belong to a certain test group
+        '''
+        tests = []
+        for l in open(os.path.join(self.srcdir, 'group')).readlines():
+            m = self.GROUP_TEST_LINE_RE.match(l)
+            if m is not None:
+                test = m.groups()[0]
+                groups = m.groups()[1]
+                if group in groups.split():
+                    if test not in tests:
+                        tests.append(test)
+        return tests
+
+    def test(self):
+
+        os.chdir(self.srcdir)
+        test_number = self.params.get('test_number', default='')
+        skip_dangerous = self.params.get('skip_dangerous', default=True)
+        if test_number == '000':
+            self.log.info('Dummy test to setup xfstests')
+            return
+
+        if test_number not in self._get_available_tests():
+            raise self.error('test file %s not found' % test_number)
+
+        if skip_dangerous:
+            if test_number in self._get_tests_for_group('dangerous'):
```

Comment:
Overall, this test looks good. I would move this 'skip_dangerous'  part to `setUp()`, since there we can use `self.skip()`, which is a perfect fir for this case. Notice that `self.skip()` cannot be used outside `setUp()`. Since `setUp()` is executed once per test, you can do all the parametrization there, and skip a test when applies. Please let me know if you are going to make such improvement or if you prefer to keep the test as is.